keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,1027 @@
+CIDRs (more details on this are discussed below).
+
+### Risks and Mitigations
+
+-   Racing kube-controller-managers. If multiples of the controller are running
+    (as in a HA control plane), how do they coordinate?
+    -   The controllers will coordinate using the existing
+        kube-controller-manager leader election.
+
+## Design Details
+
+### Pre-Requisites
+
+This KEP assumes that the only consumer of the `--cluster-cidr` value is the
+NodeIPAM controller. [KEP
+2450](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules)
+proposed modifications to the kube-proxy to remove it's dependence on a
+monolithic ClusterCIDR.  The kube-proxy flag `--detect-local-mode` must be set
+to `NodeCIDR` to properly handle nodes having discontiguous Pod CIDRs.
+
+Users not using kube-proxy must ensure that any components they have installed
+do not assume Kubernetes has a single continguous Pod CIDR.
+
+### New Resource
+
+```go
+type ClusterCIDRConfig struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec ClusterCIDRConfigSpec
+  Status ClusterCIDRConfigStatus
+}
+
+type ClusterCIDRConfigSpec struct {
+    // This defines which nodes the config is applicable to. A nil selector can
+    // be applied to any node.
+    // +optional
+    NodeSelector *v1.LabelSelector
+
+    // This defines the IPv4 CIDR assignable to nodes selected by this config.
+    // +optional
+    IPv4 *ClusterCIDRSpec
+
+    // This defines the IPv6 CIDR assignable to nodes selected by this config.
+    // +optional
+    IPv6 *ClusterCIDRSpec
+}
+
+type ClusterCIDRSpec struct {
+    // An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    // +required
+    CIDR string
+
+    // Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    // Users would have to ensure that the kubelet doesn't try to schedule more
+    // pods than are supported by the node's netmask (i.e. the kubelet's
+    // --max-pods flag)
+    // +required
+    PerNodeMaskSize int
+}
+
+type ClusterCIDRConfigStatus struct {
+}
+```
+
+#### Expected Behavior
+
+-   `NodeSelector`, `IPv4`, and `IPv6` are immutable after creation.
+
+-   `IPv4.PerNodeMaskSize` and `IPv6.PerNodeMaskSize` must specify the same
+    number of IP addresses:
+
+    ```32 - IPv4.PerNodeMaskSize == 128 - IPv6.PerNodeMaskSize```
+
+-   Each node will be assigned all Pod CIDRs from a matching config.
+    Consider the following example:
+    
+    ```go
+    {
+        IPv4: {
+            CIDR:            "10.0.0.0/20",
+            PerNodeMaskSize: "22",
+        },
+        IPv6: {
+            CIDR:            "fd12:3456:789a:1::/64"
+            PerNodeMaskSize: "118",
+        },
+    }
+    ```
+    Only 4  nodes may be allocated from this `ClusterCIDRConfig` as only 4 IPv4
+    Pod CIDRs can be partitioned from the IPv4 CIDR. The remaining IPv6 Pod
+    CIDRs may be used if referenced in another `ClusterCIDRConfig`.
```

Comment:
Should we allow redefinition of CIDRs (overlapping)?

Reply:
I think we should -- there's value in letting users specify a large IPv6 that they reuse across a bunch of IPv4 blocks. Otherwise, users have to start being very deliberate about manually partitioning which seems rather toilsome.